caas/kubernetes/provider/specs/builder_test.go

```diff
@@ -0,0 +1,459 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package specs_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/golang/mock/gomock"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	appsv1 "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	apischema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/rest/fake"
+
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/caas/kubernetes/provider"
+	providermocks "github.com/juju/juju/caas/kubernetes/provider/mocks"
+	k8sspecs "github.com/juju/juju/caas/kubernetes/provider/specs"
+	k8sspecsmocks "github.com/juju/juju/caas/kubernetes/provider/specs/mocks"
+	"github.com/juju/juju/cloud"
+	k8sannotations "github.com/juju/juju/core/annotations"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/testing"
+)
+
+type builderSuite struct {
+	testing.BaseSuite
+	mockRestClients map[apischema.GroupVersion]*providermocks.MockRestClientInterface
+	mockRestMapper  *k8sspecsmocks.MockRESTMapper
+
+	builder k8sspecs.DeployerInterface
+
+	namespace        string
+	deploymentParams *caas.DeploymentParams
+	labels           map[string]string
+	annotations      k8sannotations.Annotation
+}
+
+var rawK8sSpec = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app: nginx
+  namespace: ns-will-be-overwritten
+spec:
+  replicas: 3
```

Comment:
How do we handle a mismatch between this and num units the user asks for?
I think num units should override this